packages/native/src/useLinking.tsx

```diff
@@ -116,203 +306,143 @@ export default function useLinking(
-
-    return () => window.removeEventListener('popstate', onPopState);
-  }, [enabled, ref]);
+    });
+  }, [enabled, history, ref]);
 
   React.useEffect(() => {
     if (!enabled) {
       return;
     }
 
-    if (ref.current && previousStateLengthRef.current === undefined) {
-      previousStateLengthRef.current = getStateLength(
-        ref.current.getRootState()
-      );
-    }
+    if (ref.current) {
+      // We need to record the current metadata on the first render if they aren't set
+      // This will allow the initial state to be in the history entry
+      const state = ref.current.getRootState();
+      const path = getPathFromStateRef.current(state, configRef.current);
 
-    if (ref.current && location.pathname + location.search === '/') {
-      history.replaceState(
-        { index: history.state?.index ?? 0 },
-        '',
-        getPathFromStateRef.current(
-          ref.current.getRootState(),
-          configRef.current
-        )
-      );
+      if (previousStateRef.current === undefined) {
+        previousStateRef.current = state;
+      }
+
+      history.replace({ path, state });
     }
 
-    const unsubscribe = ref.current?.addListener('state', () => {
+    const onStateChange = async () => {
       const navigation = ref.current;
 
-      if (!navigation) {
+      if (!navigation || !enabled) {
         return;
       }
 
+      const previousState = previousStateRef.current;
       const state = navigation.getRootState();
-      const path = getPathFromStateRef.current(state, configRef.current);
 
-      const previousStateLength = previousStateLengthRef.current ?? 1;
-      const stateLength = getStateLength(state);
+      const pendingPath = pendingPopStatePathRef.current;
+      const path = getPathFromStateRef.current(state, configRef.current);
 
-      if (pendingStateMultiUpdateRef.current) {
-        if (location.pathname + location.search === path) {
-          pendingStateMultiUpdateRef.current = false;
-        } else {
-          return;
-        }
-      }
+      previousStateRef.current = state;
+      pendingPopStatePathRef.current = undefined;
 
-      previousStateLengthRef.current = stateLength;
+      // To detect the kind of state change, we need to:
+      // - Find the common focused navigation state in previous and current state
+      // - If only the route keys changed, compare history/routes.length to check if we go back/forward/replace
+      // - If no common focused navigation state found, it's a replace
+      const [previousFocusedState, focusedState] = findMatchingState(
+        previousState,
+        state
+      );
 
       if (
-        pendingStateUpdateRef.current &&
-        location.pathname + location.search === path
+        previousFocusedState &&
+        focusedState &&
+        // We should only handle push/pop if path changed from what was in last `popstate`
+        // Otherwise it's likely a change triggered by `popstate`
```

Comment:
can it happen?

Reply:
`popstate` calls `resetRoot` (or dispatches navigate action), so every `popstate` will trigger a state change event